### PR TITLE
Expose additional builder booster related flags in the vc

### PR DESCRIPTION
**Pull request by eserilev**
## Issue Addressed

#5039 

## Proposed Changes

Add two new flags to the VC

`prefer_builder_proposals`
`builder_boost_factor`

Use these new flags plus the existing `builder_proposals` flag to 'calculate' the correct boost factor.  

Enable these flags to be set via cli and relevant validator management api endpoints

### CODE REVIEW

File: book/src/help_vc.md
Comment (paulhauner) [inline book/src/help_vc.md]: Thanks for getting this ready so quick @eserilev 🙏 

My preference is to rename this to `--always-prefer-builder-proposals`. I don't love the "always" prefix, but that's the name of the flag that was removed. I think it would be nice to avoid users needing to remove `always-` from their configs in order to update.

I'd be open to leaving this flag and adding an `--always..` alias with a deprecation warning if we really want to drop the "always". I just think it'll be a bit of a bumpy upgrade if we drop the flag all of a sudden. We know at all RP users will have this flag, so it's commonly used.

Comment (michaelsproul) [reply book/src/help_vc.md]: The previous flag was a BN flag and the new one is a VC flag, so there's no backwards-compat concern re: keeping the flag name the same

RocketPool gang will need to use a VC flag instead of a BN flag

Comment (paulhauner) [reply book/src/help_vc.md]: Oh crap, sorry!

File: validator_client/src/validator_store.rs
Comment (paulhauner) [inline validator_client/src/validator_store.rs]: Can we remove this field?

File: validator_client/src/config.rs
Comment (paulhauner) [inline validator_client/src/config.rs]:
```suggestion
        config.builder_boost_factor = parse_optional(cli_args, "builder-boost-factor")?;
```

File: validator_client/src/block_service.rs
Comment (paulhauner) [inline validator_client/src/block_service.rs]:
```suggestion
        if prefer_builder_proposals {
            return Some(u64::MAX);
        }

        if let Some(builder_boost_factor) = builder_boost_factor {
            // if builder boost factor is set to 100 it should be treated
            // as None to prevent unnecessary calculations that could
            // lead to loss of information.
            if builder_boost_factor != 100 {
                return Some(builder_boost_factor);
            }
        }
```

I'm not sure if there's an objective truth behind how these preferences should work, but I was thinking that it might be easier for debugging this way.

I imagine a confused config that has `prefer_builder_payloads == True` and `builder_boost_factor == 0`. I feel like it would be easier to answer the question "why is this still preferring builder payloads?" rather than "why is it preferring local payloads?".

**Review by paulhauner — COMMENTED** [review]
Looking great! I just a have a few minor suggestions. There are also some docs over in the Lighthouse book that might be nice to update, eg https://github.com/sigp/lighthouse/blob/stable/book/src/api-vc-endpoints.md#patch-lighthousevalidatorsvoting_pubkey

**Comment by eserilev** [comment]
thanks for the review Paul. I've made changes based on your feedback and added some additional documentation to the Lighthouse book.

File: validator_client/src/block_service.rs
Comment (realbigsean) [inline validator_client/src/block_service.rs]: The order here looks like if the `--builder-proposals` flag is forgotten, the `--prefer-builder-proposals` and `--builder-boost-factor` flags don't have any effect. Maybe we should use `requires` in `clap` to make sure `--builder-proposals` is provided with the other two. Alternatively, `--builder-proposals` could be implied by the presence of one of the other two flags.

Comment (eserilev) [reply validator_client/src/block_service.rs]: Yeah thats a good point. I think, at the very least, `--builder-proposals` should be implied by the presence of one of the other two flags. Reason being is the validator management api and validator yml definition files wont trigger the `requires` in `clap`.

**Review by paulhauner — APPROVED** [review]
Nice! Once we have approval from @realbigsean as well we can merge!

**Review by realbigsean — APPROVED** [review]
Awesome!

File: validator_client/src/block_service.rs
Comment (michaelsproul) [inline validator_client/src/block_service.rs]:
```suggestion
    /// Translate `builder_proposals`, `builder_boost_factor` and `prefer_builder_proposals`` to a
```

File: validator_client/src/validator_store.rs
Comment (michaelsproul) [inline validator_client/src/validator_store.rs]: I think this function is missing the fallback to the process-level configuration?

It just calls `InitializedValidators::builder_boost_factor` which doesn't use the process-level config either: https://github.com/sigp/lighthouse/blob/c852a51302a256d9c305fccfb3e4eb94c6c4b5ad/validator_client/src/initialized_validators.rs#L824-L828

**Review by michaelsproul — CHANGES_REQUESTED** [review]

File: validator_client/src/validator_store.rs
Comment (jimmygchen) [reply validator_client/src/validator_store.rs]: Looking at the current logic, it looks like the process wide `prefer_builder_proposals` value would override per validator `builder_boost_factor`. It probably makes more sense that per-validator config overrides CLI flags. I've drafted some changes here:
https://github.com/jimmygchen/lighthouse/commit/03152d20f3951fd4cca1c2e35a73c06be1cb0bb4

Going to do a bit of tidy up now with some tests, and will push the changes to this branch.

File: validator_client/src/block_service.rs
Comment (jimmygchen) [reply validator_client/src/block_service.rs]: Updated in https://github.com/sigp/lighthouse/pull/5086/commits/b225328f9dd6a0bcfcfe83229b58d8774f9771d5

File: validator_client/src/validator_store.rs
Comment (jimmygchen) [reply validator_client/src/validator_store.rs]: Updated in https://github.com/sigp/lighthouse/pull/5086/commits/b225328f9dd6a0bcfcfe83229b58d8774f9771d5

Per-validator config is now prioritised over process level configurations. I'm looking at potentially adding an http test for testing this logic, but may require a bit of refactor of the tests.

Comment (jimmygchen) [reply validator_client/src/validator_store.rs]: Testing all the permutations is a bit tricky with the current setup - we'll need to run a `ProductionValidatorClient` or make `BlockService` a bit more testable.

We could potentially introduce some mocking libraries, or add some abstractions to improve the validator client testability - @michaelsproul and I chatted about this a while ago, but I couldn't find where our discussion was.

There's a bit of setup required for this so might have to leave it out of this PR.

**Comment by jimmygchen** [comment]
I've added a new test (locally) that covers the scenario where process defaults is present, and checking that the per-validator value overrides them - the test is failing though, I'm investigating now.

**Comment by eserilev** [comment]
thanks for helping get this across the finish line everyone!

**Comment by jimmygchen** [comment]
I've added some more tests and it's looking good now after Paul's fixes! 
Perhaps another approval from @realbigsean again and we can merge?

**Review by realbigsean — APPROVED** [review]
new changes look good!

**Review by michaelsproul — APPROVED** [review]
LGTM, let's ship :rocket:

**Pull request merged**
